src/distributions/weighted/mod.rs

```diff
@@ -125,9 +126,144 @@ impl<X: SampleUniform + PartialOrd> WeightedIndex<X> {
             return Err(WeightedError::AllWeightsZero);
         }
-        let distr = X::Sampler::new(zero, total_weight);
+        let distr = X::Sampler::new(zero, total_weight.clone());
 
-        Ok(WeightedIndex { cumulative_weights: weights, weight_distribution: distr })
+        Ok(WeightedIndex { cumulative_weights: weights, total_weight, weight_distribution: distr })
+    }
+
+    /// Update a subset of weights, without changing the number of weights.
+    ///
+    /// `new_weights` must be sorted by the index.
+    ///
+    /// Using this method instead of `new` might be more efficient if only a small number of
+    /// weights is modified. No allocations are performed, unless the weight type `X` uses
+    /// allocation internally.
+    ///
+    /// In case of error, `self` is not modified.
+    pub fn update_weights(&mut self, new_weights: &[(usize, &X)]) -> Result<(), WeightedError>
+        where X: for<'a> ::core::ops::AddAssign<&'a X> +
+                 for<'a> ::core::ops::SubAssign<&'a X> +
+                 Clone +
+                 Default {
+        if new_weights.is_empty() {
+            return Ok(());
+        }
+
+        let zero = <X as Default>::default();
+
+        let mut total_weight = self.total_weight.clone();
+
+        // Check for errors first, so we don't modify `self` in case something
+        // goes wrong.
+        let mut prev_i = None;
+        for &(i, w) in new_weights {
+            if let Some(old_i) = prev_i {
+                if old_i >= i {
+                    return Err(WeightedError::InvalidWeight);
+                }
+            }
+            if *w < zero {
+                return Err(WeightedError::InvalidWeight);
+            }
+            if i >= self.cumulative_weights.len() + 1 {
+                return Err(WeightedError::TooMany);
+            }
+
+            // Unfortunately, we will have to calculate the non-cumulative
+            // weight a second time, to avoid producing an invalid state of
+            // `self`.
+            let mut old_w = if i < self.cumulative_weights.len() {
+                self.cumulative_weights[i].clone()
+            } else {
+                self.total_weight.clone()
+            };
+            if i > 0 {
+                old_w -= &self.cumulative_weights[i - 1];
+            }
+
+            total_weight -= &old_w;
+            total_weight += w;
+            prev_i = Some(i);
+        }
+        if total_weight == zero {
+            return Err(WeightedError::AllWeightsZero);
+        }
+
+        // Update the weights. Because we checked all the preconditions in the
+        // previous loop, this should never panic.
+        let mut iter = new_weights.iter();
+        let &(first_new_index, first_new_weight) = iter.next().unwrap();
+
+        // `X` might be an unsigned type, so we have to be careful to avoid
+        // negative numbers. This is done by tracking the sign of `change` using
+        // `pos_sign`.
```

Comment:
There is a way around this: ensure you are never dealing with negative values! Instead of keeping a delta, re-calculate the cumulative weights as you go:
```
if i == next.0 {
    cumulative_weight += next.1;
} else {
    let mut tmp = cumulative_weights[i].clone();
    tmp -= prev_weight;  // we know this is positive
    cumulative_weight += tmp;
}
// to save a clone, we do a swap:
prev_weight = cumulative_weight.clone();
mem::swap(&mut prev_weight, &mut cumulative_weights[i]);
```
Before the loop, initialise `cumulative_weight = if next.0 > 0 { cumulative_weights[next.0 - 1] } else 0 };`

Reply:
Done.